### ISSUE TITLE
Broken pipe

### ISSUE THREAD

**Issue by Tommassino** [polynote/polynote#597]
Randomly when using polynote I get a broken pipe and the kernel dies. i.e. it happened multiple times, and I could not determine any obvious behavior that would cause it. 

in the UI i get the following exception
```
java.io.IOException: Broken pipe

Uncaught exception: Broken pipe (java.io.IOException)
sun.nio.ch.FileDispatcherImpl.write0(FileDispatcherImpl.java:-2)
sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:93)
sun.nio.ch.IOUtil.write(IOUtil.java:65)
sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:471)
polynote.kernel.remote.SocketTransport$FramedSocket$$anonfun$write$1.apply$mcV$sp(transport.scala:386)
polynote.kernel.remote.SocketTransport$FramedSocket$$anonfun$write$1.apply(transport.scala:375)
polynote.kernel.remote.SocketTransport$FramedSocket$$anonfun$write$1.apply(transport.scala:375)
zio.blocking.Blocking$Service$$anonfun$effectBlocking$1$$anonfun$apply$3.apply(Blocking.scala:133)
zio.blocking.Blocking$Service$$anonfun$effectBlocking$1$$anonfun$apply$3.apply(Blocking.scala:127)
zio.internal.FiberContext.evaluateNow(FiberContext.scala:333)
zio.internal.FiberContext.zio$internal$FiberContext$$run$body$2(FiberContext.scala:602)
zio.internal.FiberContext$$anonfun$7.run(FiberContext.scala:602)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
java.lang.Thread.run(Thread.java:748)
```

in the log (where i think its relevant):
```
... these repeat a couple of times (might not be related, its happening a LOT)
Oct 31 15:03:14  bash[29136]: [ERROR] Fiber failed.
Oct 31 15:03:14  bash[29136]:         A checked error was not handled.
Oct 31 15:03:14  bash[29136]:         jep.JepException: <class 'TypeError'>: descriptor '__repr__' of 'object' object needs an argument
Oct 31 15:03:14  bash[29136]:                 at jep.python.PyCallable.call(Native Method)
Oct 31 15:03:14  bash[29136]:                 at jep.python.PyCallable.callAs(PyCallable.java:110)
Oct 31 15:03:14  bash[29136]:                 at polynote.runtime.python.PythonObject$$anonfun$1.apply(PythonObject.scala:25)
Oct 31 15:03:14  bash[29136]:                 at polynote.runtime.python.PythonObject$$anonfun$1.apply(PythonObject.scala:22)
Oct 31 15:03:14  bash[29136]:                 at zio.blocking.Blocking$Service$$anonfun$effectBlocking$1$$anonfun$apply$3.apply(Blocking.scala:133)
Oct 31 15:03:14  bash[29136]:                 at zio.blocking.Blocking$Service$$anonfun$effectBlocking$1$$anonfun$apply$3.apply(Blocking.scala:127)
Oct 31 15:03:14  bash[29136]:                 at zio.internal.FiberContext.evaluateNow(FiberContext.scala:333)
Oct 31 15:03:14  bash[29136]:                 at zio.internal.FiberContext.zio$internal$FiberContext$$run$body$2(FiberContext.scala:602)
Oct 31 15:03:14  bash[29136]:                 at zio.internal.FiberContext$$anonfun$7.run(FiberContext.scala:602)
Oct 31 15:03:14  bash[29136]:                 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
Oct 31 15:03:14  bash[29136]:                 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
Oct 31 15:03:14  bash[29136]:                 at java.lang.Thread.run(Thread.java:748)
...
... first time a broken pipe is mentioned
Oct 31 15:07:06  bash[29136]: [INFO]  Response stream terminated
Oct 31 15:07:07  bash[29136]: [ERROR] Fiber failed.
Oct 31 15:07:07  bash[29136]:         A checked error was not handled.
Oct 31 15:07:07  bash[29136]:         java.io.IOException: Broken pipe
Oct 31 15:07:07  bash[29136]:                 at sun.nio.ch.FileDispatcherImpl.write0(Native Method)
Oct 31 15:07:07  bash[29136]:                 at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
Oct 31 15:07:07  bash[29136]:                 at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:93)
Oct 31 15:07:07  bash[29136]:                 at sun.nio.ch.IOUtil.write(IOUtil.java:65)
Oct 31 15:07:07  bash[29136]:                 at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:471)
Oct 31 15:07:07  bash[29136]:                 at polynote.kernel.remote.SocketTransport$FramedSocket$$anonfun$write$1.apply$mcV$sp(transport.scala:383)
Oct 31 15:07:07  bash[29136]:                 at polynote.kernel.remote.SocketTransport$FramedSocket$$anonfun$write$1.apply(transport.scala:375)
Oct 31 15:07:07  bash[29136]:                 at polynote.kernel.remote.SocketTransport$FramedSocket$$anonfun$write$1.apply(transport.scala:375)
Oct 31 15:07:07  bash[29136]:                 at zio.blocking.Blocking$Service$$anonfun$effectBlocking$1$$anonfun$apply$3.apply(Blocking.scala:133)
Oct 31 15:07:07  bash[29136]:                 at zio.blocking.Blocking$Service$$anonfun$effectBlocking$1$$anonfun$apply$3.apply(Blocking.scala:127)
Oct 31 15:07:07  bash[29136]:                 at zio.internal.FiberContext.evaluateNow(FiberContext.scala:333)
Oct 31 15:07:07  bash[29136]:                 at zio.internal.FiberContext.zio$internal$FiberContext$$run$body$2(FiberContext.scala:602)
Oct 31 15:07:07  bash[29136]:                 at zio.internal.FiberContext$$anonfun$7.run(FiberContext.scala:602)
Oct 31 15:07:07  bash[29136]:                 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
Oct 31 15:07:07  bash[29136]:                 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
Oct 31 15:07:07  bash[29136]:                 at java.lang.Thread.run(Thread.java:748)
Oct 31 15:07:07  bash[29136]:
Oct 31 15:07:07  bash[29136]:         Fiber:605 was supposed to continue to:
Oct 31 15:07:07  bash[29136]:           a future continuation at fs2.internal.Algebra$$anonfun$compileLoop$1.apply(Algebra.scala:349)
Oct 31 15:07:07  bash[29136]:           a future continuation at fs2.internal.Algebra$$anonfun$compile$1.apply(Algebra.scala:167)
Oct 31 15:07:07  bash[29136]:           a future continuation at zio.ZIO$$anonfun$run$2.apply(ZIO.scala:1120)
Oct 31 15:07:07  bash[29136]:           a future continuation at zio.interop.CatsEffect$$anonfun$bracketCase$1.apply(cats.scala:223)
Oct 31 15:07:07  bash[29136]:           a future continuation at fs2.Stream$CompileOps$$anonfun$foldChunks$2.apply(Stream.scala:4023)
Oct 31 15:07:07  bash[29136]:
Oct 31 15:07:07  bash[29136]:         Fiber:605 execution trace:
Oct 31 15:07:07  bash[29136]:           at fs2.internal.Algebra$$anonfun$fs2$internal$Algebra$$closeAndGo$1$1$$anonfun$apply$10.apply(Algebra.scala:306)
Oct 31 15:07:07  bash[29136]:           at fs2.internal.CompileScope$$anonfun$openAncestor$2$$anonfun$apply$24.apply(CompileScope.scala:232)
Oct 31 15:07:07  bash[29136]:           at cats.effect.concurrent.Ref$SyncRef$$anonfun$get$1.apply(Ref.scala:221)
Oct 31 15:07:07  bash[29136]:           at fs2.internal.Algebra$$anonfun$fs2$internal$Algebra$$closeAndGo$1$1.apply(Algebra.scala:305)
Oct 31 15:07:07  bash[29136]:           at fs2.internal.CompileScope$$anonfun$close$2$$anonfun$apply$15$$anonfun$apply$17$$anonfun$apply$20$$anonfun$apply$23.apply(CompileScope.scala:218)
Oct 31 15:07:07  bash[29136]:           at cats.effect.concurrent.Ref$SyncRef$$anonfun$modify$1.apply(Ref.scala:255)
Oct 31 15:07:07  bash[29136]:           at fs2.internal.CompileScope$$anonfun$close$2$$anonfun$apply$15$$anonfun$apply$17$$anonfun$apply$20.apply(CompileScope.scala:217)
Oct 31 15:07:07  bash[29136]:           at fs2.internal.CompileScope$$anonfun$close$2$$anonfun$apply$15$$anonfun$apply$17.apply(CompileScope.scala:216)
Oct 31 15:07:07  bash[29136]:           at fs2.internal.CompileScope$$anonfun$fs2$internal$CompileScope$$traverseError$1.apply(CompileScope.scala:193)
Oct 31 15:07:07  bash[29136]:           at cats.data.ChainInstances$$anon$1$$anonfun$traverse$1$$anonfun$apply$3.apply(Chain.scala:715)
Oct 31 15:07:07  bash[29136]:           at zio.ZIO$$anonfun$zipWith$1.apply(ZIO.scala:1490)
Oct 31 15:07:07  bash[29136]:           at zio.ZIO$$anonfun$12.apply(ZIO.scala:2661)
Oct 31 15:07:07  bash[29136]:           at zio.ZIO$$anonfun$unit$1.apply(ZIO.scala:1405)
Oct 31 15:07:07  bash[29136]:           at zio.internal.FiberContext$$anonfun$interrupt$1.apply(FiberContext.scala:614)
Oct 31 15:07:07  bash[29136]:           at zio.RIO$$anonfun$effectSuspend$1.apply(RIO.scala:207)
Oct 31 15:07:07  bash[29136]:           at fs2.internal.Resource$$anon$1$$anonfun$release$2.apply(Resource.scala:120)
Oct 31 15:07:07  bash[29136]:           at cats.effect.concurrent.Ref$SyncRef$$anonfun$modify$1.apply(Ref.scala:255)
...
```

Its pretty serious for me, since the random restart makes it pretty annoying to use. I hope I provided enough information.

**Comment by Tommassino** [comment]
ah, bummer, i dont have journalctl set up to retain logs between reboots, so i do not have the full log anymore. If it happens again ill send the full log.

**Comment by Tommassino** [comment]
Here is a full log. The broken pipe happened when i was working in python, plotting some stuff etc.

[polynote.log](https://github.com/polynote/polynote/files/3797348/polynote.log)

**Comment by jonathanindig** [comment]
Hi @Tommassino are you able to provide your notebook file? Also, any details on your system setup could be useful as well, thanks!

**Comment by jeremyrsmith** [comment]
I believe this should be largely solved by #629 – once we release that I'll close this issue.

**Comment by jonathanindig** [comment]
Resolved by #629 (released as part of 0.2.13)
